lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx

```diff
@@ -294,65 +304,109 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
+        mustInclude = false;
+      }
+      else
+      {
+        for (UnsignedInteger variableIndex = 0; variableIndex < inputDimension; ++variableIndex)
         {
-          totalVariance += coefficientI * coefficientI;
-          // Set the exponents corresponding to the group to zero
-          for (UnsignedInteger j = 0; j < orderSobolIndice; ++j)
-            multiIndices[variableIndices[j]] = 0;
-          // Now check that all the indices are zero
-          if (*std::max_element(multiIndices.begin(), multiIndices.end()) == 0)
-            covarianceVariables += coefficientI * coefficientI;
-        } // *std::max_element(multiIndices.begin(), multiIndices.end()) > 0
-      } // if coefficientI <> 0
+          // Check that each variable having a nonzero degree is in the group
+          if (multiIndices[variableIndex] > 0 and not variablesGroup.contains(variableIndex)) 
+          {
+            mustInclude = false;
+            break;
+          }
+        }
+      }
+      if (mustInclude) covarianceVariables += coefficientI * coefficientI;
     } // coefficientIndices[i] > 0
   } // Loop over the coefficients
   if (totalVariance > 0.0) return covarianceVariables / totalVariance;
   else return 0.0;
 }
 
 /* Sobol grouped total index accessor */
-Scalar FunctionalChaosSobolIndices::getSobolGroupedTotalIndex(const Indices & variableIndices,
+Scalar FunctionalChaosSobolIndices::getSobolGroupedTotalIndex(const Indices & variablesGroup,
     const UnsignedInteger marginalIndex) const
 {
-  // Compute the complementary indice
   const UnsignedInteger inputDimension = functionalChaosResult_.getDistribution().getDimension();
-  if (!variableIndices.check(inputDimension)) throw InvalidArgumentException(HERE) << "The variable indices of a Sobol indice must be in the range [0, dim-1] and must be different.";
-  const Indices complementaryVariableIndices = variableIndices.complement(inputDimension);
-  // Compute total index from complementary first index
-  const Scalar complementaryFirstIndex = getSobolGroupedIndex(complementaryVariableIndices, marginalIndex);
-  const Scalar groupTotalIndex = 1.0 - complementaryFirstIndex;
-  return groupTotalIndex;
+  if (!variablesGroup.check(inputDimension)) throw InvalidArgumentException(HERE) << "The variable indices of a Sobol' index must be in the range [0, dim-1] and must be different.";
+  const Sample coefficients(functionalChaosResult_.getCoefficients().getMarginal(marginalIndex));
+  const Indices coefficientIndices(functionalChaosResult_.getIndices());
+  const UnsignedInteger size = coefficients.getSize();
+  Scalar covarianceVariables = 0.0;
+  const EnumerateFunction enumerateFunction(functionalChaosResult_.getOrthogonalBasis().getEnumerateFunction());
+  // Sum the contributions of all the coefficients associated to a basis vector involving only the needed variables
+  Scalar totalVariance = 0.0;
+  const UnsignedInteger groupDimension = variablesGroup.getSize();
+  bool mustInclude;
```

Comment:
we could initialize it once to false

Reply:
Ok. Fixed at e3cd3cfe3cb97c40515f5fbf3de21a1d2300b3a9